From ea698e11316b837ca13944f013eb13446501079e Mon Sep 17 00:00:00 2001
From: Brian Silverman <bsilverman@iveia.com>
Date: Wed, 22 Apr 2020 21:26:54 -0400
Subject: [PATCH] zap: add driver

Driver currently requires allocating DMA memory outside of Linux management via
alloc_bootmem_pages().

Disable driver (and mem alloc) with kernel cmdline "nozap".
---
 drivers/char/Makefile |  1 +
 init/main.c           | 13 +++++++++++++
 2 files changed, 14 insertions(+)

diff --git a/drivers/char/Makefile b/drivers/char/Makefile
index b8d42b4e979b..3e43624ce541 100644
--- a/drivers/char/Makefile
+++ b/drivers/char/Makefile
@@ -3,6 +3,7 @@
 # Makefile for the kernel character device drivers.
 #
 
+obj-y				+= zap/
 obj-y				+= mem.o random.o
 obj-$(CONFIG_TTY_PRINTK)	+= ttyprintk.o
 obj-y				+= misc.o
diff --git a/init/main.c b/init/main.c
index 6e281285927b..a269a4cd7aaa 100644
--- a/init/main.c
+++ b/init/main.c
@@ -529,6 +529,11 @@ static void __init mm_init(void)
 	pti_init();
 }
 
+extern volatile void * tx_vaddr;
+extern volatile void * rx_vaddr;
+bool nozap;
+core_param(nozap, nozap, bool, 0644);
+
 asmlinkage __visible void __init start_kernel(void)
 {
 	char *command_line;
@@ -580,6 +585,14 @@ asmlinkage __visible void __init start_kernel(void)
 		parse_args("Setting init args", after_dashes, NULL, 0, -1, -1,
 			   NULL, set_init_arg);
 
+    if (nozap) {
+        pr_notice("zap: SKIPPING alloc_bootmem_pages()\n");
+    } else {
+        pr_notice("zap: calling alloc_bootmem_pages()\n");
+        rx_vaddr = alloc_bootmem_pages(128*1024*1024);
+        tx_vaddr = alloc_bootmem_pages(4*1024*1024);
+    }
+
 	jump_label_init();
 
 	/*
-- 
2.17.1

